master/internal/authz/authz_provider_type.go

```diff
@@ -0,0 +1,63 @@
+package authz
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/determined-ai/determined/master/internal/config"
+	"golang.org/x/exp/maps"
+)
+
+type AuthZProviderType[T any] struct {
+	registry map[string]T
+	once     sync.Once
+}
+
+func (p *AuthZProviderType[T]) Register(authZType string, impl T) {
+	p.once.Do(func() {
+		p.registry = make(map[string]T)
+	})
+	// TODO(ilia): keep registry here or global in internal/authz/authz_basic.go
+	config.RegisterAuthZType(authZType)
+	if _, ok := p.registry[authZType]; ok {
+		panic(fmt.Errorf("can't do double register of type %s for: %s", authZType, p.string()))
+	}
+	p.registry[authZType] = impl
+}
+
+func (p *AuthZProviderType[T]) string() string {
+	return reflect.TypeOf((*T)(nil)).Elem().String()
+}
+
+func (p *AuthZProviderType[T]) Get() T {
```

Comment:
Non blocking since this is something we will need to see how plays it out

But in the case of someone testing out RBAC they are going to run it permissive but still wanting original OSS controls (especially for things like normal users being unable to change someones else's password)

This might require us to define a new interface that basically just calls the other two interfaces which might actually be the best solution here.